mage_ai/server/websocket.py

```diff
@@ -39,25 +46,63 @@ def on_message(self, raw_message):
             manager = KernelManager(**connection)
             client = manager.client()
 
-            msg_id = client.execute(code)
+            pipeline = Pipeline(pipeline_uuid, os.path.join(os.getcwd(), 'default_repo'))
+            block = pipeline.get_block(block_uuid)
+            block_output = []
+            error = None
+            if block is not None and block.type in [BlockType.DATA_LOADER, BlockType.TRANSFORMER]:
+                try:
+                    output = asyncio.run(block.execute(custom_code=code))
+                    if len(output) > 0:
+                        for out in output:
+                            if type(out) == pd.DataFrame \
+                                and out.shape[0] > DATAFRAME_OUTPUT_SAMPLE_COUNT:
 
-            WebSocketServer.running_executions_mapping.add((msg_id, block_uuid))
+                                out = out.iloc[:DATAFRAME_OUTPUT_SAMPLE_COUNT]
+                            block_output.append(out)
+                except Exception as err:
+                    error = traceback.format_exc()
+                # Run with no code because we still need to send a message
+                msg_id = client.execute('')
+            else:
+                msg_id = client.execute(code)
+
+            value = dict(
+                block_uuid=block_uuid,
+                block_output=block_output,
+                error=error,
+            )
+
+            WebSocketServer.running_executions_mapping[msg_id] = value
         elif output:
             self.send_message(output)
 
     @classmethod
     def send_message(self, message: dict) -> None:
         msg_id = message['msg_id']
-        msg_id_uuid_tuple = find(
-            lambda tup: tup[0] == msg_id,
-            list(WebSocketServer.running_executions_mapping),
+        msg_id_value = WebSocketServer.running_executions_mapping[msg_id]
+        uuid = msg_id_value['block_uuid']
+        output = msg_id_value['block_output']
+        error = msg_id_value['error']
+
+        output_dict = dict(uuid=uuid)
+        if error is not None:
+            output_dict['data'] = error
+            output_dict['msg_type'] = 'error'
+        elif len(output) > 0:
+            df = find(lambda val: type(val) == pd.DataFrame, output)
+            output_dict['data'] = [
+                df.columns.to_list(),
```

Comment:
is this format used by frontend now? otherwise, can we use this format for consistency? https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/models/block.py#L239-L242

Reply:
it's not used in the frontend right now, but tommy asked for it in this format. @tommydangerous is it ok if we change it to the above format for consistency?